File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_search/P875_KoKoEatingBananas.java
package com.dsa.leetcode.labuladong.binary_search;

public class P875_KoKoEatingBananas {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int right = Integer.MIN_VALUE;
        for (int pile : piles) {
            right = Math.max(right, pile);
        }
        while (left < right) {
            int mid = left + (right - left) / 2;
            int target = calculate(piles, mid);
            if (target <= h) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    private int calculate(int[] piles, int mid) {
        int h = 0;
        for (int pile : piles) {
            h += pile / mid;
            if (pile % mid != 0) {
                h++;
            }
        }
        return h;
    }

    public static void main(String[] args) {
        int[] piles = {312884470};
        int h = 312884469;
        P875_KoKoEatingBananas p875 = new P875_KoKoEatingBananas();
        int calculate = p875.minEatingSpeed(piles, h);
        System.out.println(calculate);
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_search/P875_KoKoEatingBananas.py
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = max(piles)
        while left < right:
            mid = left + (right - left) // 2
            target = self.calculate(piles, mid)
            if target <= h:
                right = mid
            else:
                left = mid + 1
        return left

    def calculate(self, piles, mid):
        h = 0
        for pile in piles:
            h += pile // mid
            if pile % mid != 0:
                h += 1
        return h


if __name__ == '__main__':
    piles = [3, 6, 7, 11]
    h = 8
    solu = Solution()
    ans = solu.minEatingSpeed(piles, h)
    print(ans)
